policies/recipes/steps.go

```diff
@@ -88,7 +95,290 @@ func parsePermissions(s string) (os.FileMode, error) {
+		fallthrough
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_GZIP:
+		fallthrough
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_BZIP:
+		fallthrough
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_LZMA:
+		fallthrough
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_XZ:
+		return extractTar(filename, step.ArchiveExtraction.Destination, step.ArchiveExtraction.GetType())
+	default:
+		return fmt.Errorf("Unrecognized archive type %q", step.ArchiveExtraction.GetType())
+	}
+}
+
+func zipIsDir(name string) bool {
+	return strings.HasSuffix(name, string(os.PathSeparator))
+}
+
+func extractZip(zipPath string, dst string) error {
+	zr, err := zip.OpenReader(zipPath)
+	if err != nil {
+		return err
+	}
+	defer zr.Close()
+
+	// Check for conflicts
+	for _, f := range zr.File {
+		filen, err := normPath(filepath.Join(dst, f.FileHeader.Name))
+		if err != nil {
+			return err
+		}
+		fmt.Printf("checking if file %s exists\n", filen)
+		stat, err := os.Stat(filen)
+		if os.IsNotExist(err) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		if zipIsDir(filen) && stat.IsDir() {
+			// it's ok if directories already exist
+			continue
+		}
+		return fmt.Errorf("file exists: %s", filen)
+	}
+
+	// Create dirs
+	for _, f := range zr.File {
+		filen, err := normPath(filepath.Join(dst, f.FileHeader.Name))
+		if err != nil {
+			return err
+		}
+
+		if !zipIsDir(filen) {
+			continue
+		}
+		_, err = os.Stat(filen)
+		if err == nil {
+			continue
+		}
+		if !os.IsNotExist(err) {
+			return err
+		}
+		err = os.MkdirAll(filen, 0755)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Create files.
+	for _, f := range zr.File {
+		filen, err := normPath(filepath.Join(dst, f.Name))
+		if err != nil {
+			return err
+		}
+		if zipIsDir(filen) {
+			continue
+		}
+		filedir := filepath.Dir(filen)
+		_, err = os.Stat(filedir)
+		if os.IsNotExist(err) {
+			err = os.MkdirAll(filedir, 0755)
+		}
+		if err != nil {
+			return err
+		}
+		fmt.Printf("os.Create %s\n", filen)
+		reader, err := f.Open()
+		if err != nil {
+			return err
+		}
+
+		dst, err := os.OpenFile(filen, os.O_RDWR|os.O_CREATE, 0755)
+		if err == nil {
+			_, err = io.Copy(dst, reader)
+		}
+		reader.Close()
+
+		if err != nil {
+			return err
+		}
+		err = os.Chtimes(filen, time.Now(), f.Modified)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func decompress(reader io.ReadCloser, compression osconfigpb.SoftwareRecipe_Step_ExtractArchive_ArchiveType) (io.ReadCloser, error) {
+	switch compression {
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR:
+		return reader, nil
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_GZIP:
+		return gzip.NewReader(reader)
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_BZIP:
+		return ioutil.NopCloser(bzip2.NewReader(reader)), nil
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_LZMA:
+		return nil, fmt.Errorf("lzma currently unsupported")
```

Comment:
Also there don't seem to be standard go libraries for lzma and xz. No third party libraries jumped out at me as particularly stable but I also didn't spend a lot of time digging on it.

Reply:
I added https://github.com/ulikunitz/xz as that seems like the most stable but I haven't dug into it too deeply